File: packages/protocol_engine/psychometrics.py
"""Psychometric fit and signal-detection measures (Python mirror).

Mirror of ``apps/flutter_app/lib/core/psychometrics.dart`` — the logistic
psychometric-function fit (deterministic coarse-to-fine ML grid search),
m-AFC d' (Hacker & Ratcliff, 1979), and the intraclass correlation
coefficients ICC(2,1) / ICC(3,1) for test–retest reliability (Shrout &
Fleiss, 1979; McGraw & Wong, 1996). Kept dependency-free (no numpy/scipy) so
the two implementations stay comparable line-for-line and are pinned by shared
golden values (``tests/test_psychometrics.py`` and the Dart harness).

Research summaries — never a diagnosis.
"""
from __future__ import annotations

import math
from dataclasses import dataclass
from typing import List, Optional, Sequence


def normal_pdf(z: float) -> float:
    return math.exp(-0.5 * z * z) / math.sqrt(2 * math.pi)


def normal_cdf(z: float) -> float:
    """Standard normal CDF via Abramowitz & Stegun 7.1.26 (|err| < 1.5e-7).

    Deliberately the same approximation as the Dart mirror (not math.erf) so
    both sides produce bit-comparable values.
    """
    t = z / math.sqrt(2)
    x = abs(t)
    u = 1 / (1 + 0.3275911 * x)
    poly = u * (
        0.254829592
        + u * (-0.284496736 + u * (1.421413741 + u * (-1.453152027 + u * 1.061405429)))
    )
    erf = 1 - poly * math.exp(-x * x)
    signed = -erf if t < 0 else erf
    return 0.5 * (1 + signed)


def pc_from_dprime(d_prime: float, m: int) -> float:
    """Proportion correct for an unbiased m-AFC observer with sensitivity d'."""
    if m < 2:
        raise ValueError(f"m-AFC needs m >= 2, got {m}")
    lo, hi, n = -8.0, 8.0, 1600  # Simpson's rule, step 0.01
    h = (hi - lo) / n

    def f(z: float) -> float:
        return normal_pdf(z - d_prime) * normal_cdf(z) ** (m - 1)

    total = f(lo) + f(hi)
    for i in range(1, n):
        total += f(lo + i * h) * (2 if i % 2 == 0 else 4)
    return min(1.0, max(0.0, total * h / 3))


def dprime_from_pc(pc: float, m: int) -> Optional[float]:
    """Sensitivity d' from proportion correct on an m-AFC task (bisection)."""
    chance = 1 / m
    if pc <= chance + 1e-3 or pc >= 0.999:
        return None
    lo, hi = 0.0, 10.0
    for _ in range(60):
        mid = (lo + hi) / 2
        if pc_from_dprime(mid, m) < pc:
            lo = mid
        else:
            hi = mid
    return (lo + hi) / 2


@dataclass(frozen=True)
class LogisticFit:
    alpha: float
    beta: float
    gamma: float
    lambda_: float
    target_proportion: float
    threshold: Optional[float]
    n_trials: int

    def probability_at(self, x: float) -> float:
        return self.gamma + (1 - self.gamma - self.lambda_) / (
            1 + math.exp(-self.beta * (x - self.alpha))
        )


def _neg_log_likelihood(
    x: Sequence[float],
    correct: Sequence[bool],
    alpha: float,
    beta: float,
    gamma: float,
    lambda_: float,
) -> float:
    nll = 0.0
    for xi, ci in zip(x, correct):
        p = gamma + (1 - gamma - lambda_) / (1 + math.exp(-beta * (xi - alpha)))
        p = min(1 - 1e-6, max(1e-6, p))
        nll -= math.log(p) if ci else math.log(1 - p)
    return nll


def fit_logistic(
    x: List[float],
    correct: List[bool],
    *,
    guess_rate: float,
    lapse_rate: float = 0.02,
    target_proportion: float = 0.707,
) -> Optional[LogisticFit]:
    """Deterministic coarse-to-fine ML fit; mirrors ``fitLogistic`` in Dart."""
    if len(x) != len(correct) or len(x) < 8:
        return None
    x_min, x_max = min(x), max(x)
    rng = x_max - x_min
    if rng <= 0:
        return None
    if not any(correct) or all(correct):
        return None

    best_alpha = (x_min + x_max) / 2
    best_beta = 2 / rng
    best_nll = math.inf

    alpha_lo, alpha_hi = x_min - rng * 0.25, x_max + rng * 0.25
    beta_lo, beta_hi = 0.1 / rng, 50 / rng
    for _round in range(3):
        steps = 24
        log_beta_lo, log_beta_hi = math.log(beta_lo), math.log(beta_hi)
        for i in range(steps + 1):
            a = alpha_lo + (alpha_hi - alpha_lo) * i / steps
            for j in range(steps + 1):
                b = math.exp(log_beta_lo + (log_beta_hi - log_beta_lo) * j / steps)
                nll = _neg_log_likelihood(x, correct, a, b, guess_rate, lapse_rate)
                if nll < best_nll:
                    best_nll, best_alpha, best_beta = nll, a, b
        a_span = (alpha_hi - alpha_lo) / steps * 2
        alpha_lo, alpha_hi = best_alpha - a_span, best_alpha + a_span
        beta_lo, beta_hi = best_beta / 2.5, best_beta * 2.5

    threshold: Optional[float] = None
    sigma = (target_proportion - guess_rate) / (1 - guess_rate - lapse_rate)
    if 0 < sigma < 1:
        threshold = best_alpha + math.log(sigma / (1 - sigma)) / best_beta

    return LogisticFit(
        alpha=best_alpha,
        beta=best_beta,
        gamma=guess_rate,
        lambda_=lapse_rate,
        target_proportion=target_proportion,
        threshold=threshold,
        n_trials=len(x),
    )


def _icc_mean_squares(
    data: Sequence[Sequence[float]],
) -> Optional[tuple]:
    """Two-way ANOVA mean squares (MSR, MSC, MSE) for an n×k score matrix.

    Returns None when the matrix is smaller than 2×2 or ragged.
    """
    n = len(data)
    if n < 2:
        return None
    k = len(data[0])
    if k < 2:
        return None
    if any(len(row) != k for row in data):
        return None
    grand = sum(v for row in data for v in row) / (n * k)
    row_means = [sum(row) / k for row in data]
    col_means = [sum(data[i][j] for i in range(n)) / n for j in range(k)]
    ssr = sum((m - grand) ** 2 for m in row_means)
    ssc = sum((m - grand) ** 2 for m in col_means)
    sse = sum(
        (data[i][j] - row_means[i] - col_means[j] + grand) ** 2
        for i in range(n)
        for j in range(k)
    )
    msr = k * ssr / (n - 1)
    msc = n * ssc / (k - 1)
    mse = sse / ((n - 1) * (k - 1))
    return msr, msc, mse


def icc21(data: Sequence[Sequence[float]]) -> Optional[float]:
    """ICC(2,1): two-way random effects, absolute agreement, single measure.

    ``data`` is an n-subjects × k-occasions matrix (each row one subject's
    scores on the same test, in occasion order). Returns None for a matrix
    smaller than 2×2, ragged rows, or zero variance.
    """
    ms = _icc_mean_squares(data)
    if ms is None:
        return None
    msr, msc, mse = ms
    n, k = len(data), len(data[0])
    denom = msr + (k - 1) * mse + k * (msc - mse) / n
    if abs(denom) < 1e-12:
        return None
    return (msr - mse) / denom


def icc31(data: Sequence[Sequence[float]]) -> Optional[float]:
    """ICC(3,1): two-way mixed effects, consistency, single measure.

    Same input and degeneracy rules as :func:`icc21`.
    """
    ms = _icc_mean_squares(data)
    if ms is None:
        return None
    msr, _msc, mse = ms
    k = len(data[0])
    denom = msr + (k - 1) * mse
    if abs(denom) < 1e-12:
        return None
    return (msr - mse) / denom

File: apps/flutter_app/lib/core/psychometrics.dart
/// Psychometric-function fit and signal-detection measures (pure Dart).
///
/// Two research-standard analyses computed from a session's per-trial data:
///
/// 1. **Logistic psychometric fit** — maximum-likelihood fit of
///    `P(correct | x) = γ + (1 − γ − λ)·σ(β(x − α))` to the (parameter,
///    correct) pairs of an adaptive run, with the guess rate γ fixed at the
///    task's chance level (1/n for nAFC) and a small fixed lapse rate λ.
///    Reports the threshold (the parameter value at the staircase's target
///    proportion), the slope β, and the fitted curve. The fit is a
///    deterministic coarse-to-fine grid search (no RNG), so results are
///    reproducible bit-for-bit.
///
/// 2. **d′ for n-alternative forced choice** — sensitivity from proportion
///    correct via the standard m-AFC observer model
///    `pc(d′) = ∫ φ(z − d′)·Φ(z)^(m−1) dz` (Hacker & Ratcliff, 1979; Green &
///    Swets, 1966), inverted numerically. For m = 2 this reduces to
///    `pc = Φ(d′/√2)`.
///
/// 3. **Intraclass correlation coefficients** — test–retest reliability from
///    an n-subjects × k-occasions score matrix via the two-way ANOVA mean
///    squares (Shrout & Fleiss, 1979; McGraw & Wong, 1996): ICC(2,1)
///    (two-way random effects, absolute agreement, single measurement) and
///    ICC(3,1) (two-way mixed effects, consistency, single measurement).
///
/// Mirrored in `packages/protocol_engine/psychometrics.py` and pinned by
/// shared golden values (`tool/verify/psychometrics_harness.dart`,
/// `tests/test_psychometrics.py`). Research summaries — never a diagnosis.
library;

import 'dart:math' as math;

/// Standard normal probability density.
double normalPdf(double z) =>
    math.exp(-0.5 * z * z) / math.sqrt(2 * math.pi);

/// Standard normal CDF via the Abramowitz & Stegun 7.1.26 erf approximation
/// (|error| < 1.5e-7) — deterministic and dependency-free.
double normalCdf(double z) {
  final t = z / math.sqrt2;
  final x = t.abs();
  final u = 1 / (1 + 0.3275911 * x);
  final poly = u *
      (0.254829592 +
          u *
              (-0.284496736 +
                  u * (1.421413741 + u * (-1.453152027 + u * 1.061405429))));
  final erf = 1 - poly * math.exp(-x * x);
  final signed = t < 0 ? -erf : erf;
  return 0.5 * (1 + signed);
}

/// Proportion correct for an unbiased m-AFC observer with sensitivity
/// [dPrime]: `∫ φ(z − d′)·Φ(z)^(m−1) dz` (Simpson's rule over z ∈ [−8, 8]).
double pcFromDPrime(double dPrime, int m) {
  if (m < 2) throw ArgumentError('m-AFC needs m >= 2, got $m');
  const lo = -8.0, hi = 8.0;
  const n = 1600; // even; step 0.01
  const h = (hi - lo) / n;
  double f(double z) =>
      normalPdf(z - dPrime) * math.pow(normalCdf(z), m - 1).toDouble();
  var sum = f(lo) + f(hi);
  for (var i = 1; i < n; i++) {
    sum += f(lo + i * h) * (i.isEven ? 2 : 4);
  }
  return (sum * h / 3).clamp(0.0, 1.0);
}

/// Sensitivity d′ from proportion correct [pc] on an m-AFC task, by bisection
/// on [pcFromDPrime]. Returns null when [pc] is at/below chance (1/m) or too
/// close to 1 for a stable estimate.
double? dPrimeFromPc(double pc, int m) {
  final chance = 1 / m;
  if (pc <= chance + 1e-3 || pc >= 0.999) return null;
  var lo = 0.0, hi = 10.0;
  for (var i = 0; i < 60; i++) {
    final mid = (lo + hi) / 2;
    if (pcFromDPrime(mid, m) < pc) {
      lo = mid;
    } else {
      hi = mid;
    }
  }
  return (lo + hi) / 2;
}

/// The result of a logistic psychometric-function fit.
class LogisticFit {
  const LogisticFit({
    required this.alpha,
    required this.beta,
    required this.gamma,
    required this.lambda,
    required this.targetProportion,
    required this.threshold,
    required this.nTrials,
  });

  /// Inflection point (parameter value at the halfway proportion).
  final double alpha;

  /// Slope (per parameter unit; positive = higher parameter is easier).
  final double beta;

  /// Guess rate (chance floor, fixed at 1/n for an nAFC task).
  final double gamma;

  /// Fixed lapse rate.
  final double lambda;

  /// The proportion the threshold is read at (e.g. 0.707 for 2-down/1-up).
  final double targetProportion;

  /// Parameter value where the fitted curve crosses [targetProportion], or
  /// null when that proportion is outside the curve's asymptotes.
  final double? threshold;

  final int nTrials;

  /// The fitted proportion correct at parameter value [x].
  double probabilityAt(double x) =>
      gamma + (1 - gamma - lambda) / (1 + math.exp(-beta * (x - alpha)));
}

double _negLogLikelihood(
  List<double> x,
  List<bool> correct,
  double alpha,
  double beta,
  double gamma,
  double lambda,
) {
  var nll = 0.0;
  for (var i = 0; i < x.length; i++) {
    var p = gamma + (1 - gamma - lambda) / (1 + math.exp(-beta * (x[i] - alpha)));
    p = p.clamp(1e-6, 1 - 1e-6);
    nll -= correct[i] ? math.log(p) : math.log(1 - p);
  }
  return nll;
}

/// Fits the logistic psychometric function to per-trial data by maximum
/// likelihood over a deterministic coarse-to-fine grid.
///
/// [x] is the adapted parameter at each trial (staircase trajectory) and
/// [correct] the per-trial outcome. [guessRate] is the task's chance level
/// (1/n for nAFC). [targetProportion] is where the threshold is read
/// (defaults to the 2-down/1-up point, ≈ 0.707). Assumes a larger parameter
/// makes the task easier — true for every staircase in this app (bigger gap /
/// delta / SNR). Returns null when there are fewer than 8 usable trials, no
/// spread in [x], or all-correct / all-wrong outcomes (nothing to fit).
LogisticFit? fitLogistic(
  List<double> x,
  List<bool> correct, {
  required double guessRate,
  double lapseRate = 0.02,
  double targetProportion = 0.707,
}) {
  if (x.length != correct.length || x.length < 8) return null;
  final xMin = x.reduce(math.min), xMax = x.reduce(math.max);
  final range = xMax - xMin;
  if (range <= 0) return null;
  final anyCorrect = correct.any((c) => c);
  final anyWrong = correct.any((c) => !c);
  if (!anyCorrect || !anyWrong) return null;

  var bestAlpha = (xMin + xMax) / 2;
  var bestBeta = 2 / range;
  var bestNll = double.infinity;

  var alphaLo = xMin - range * 0.25, alphaHi = xMax + range * 0.25;
  var betaLo = 0.1 / range, betaHi = 50 / range;
  for (var round = 0; round < 3; round++) {
    const steps = 24;
    final logBetaLo = math.log(betaLo), logBetaHi = math.log(betaHi);
    for (var i = 0; i <= steps; i++) {
      final a = alphaLo + (alphaHi - alphaLo) * i / steps;
      for (var j = 0; j <= steps; j++) {
        final b =
            math.exp(logBetaLo + (logBetaHi - logBetaLo) * j / steps);
        final nll = _negLogLikelihood(x, correct, a, b, guessRate, lapseRate);
        if (nll < bestNll) {
          bestNll = nll;
          bestAlpha = a;
          bestBeta = b;
        }
      }
    }
    // Narrow the grid around the current best for the next round.
    final aSpan = (alphaHi - alphaLo) / steps * 2;
    alphaLo = bestAlpha - aSpan;
    alphaHi = bestAlpha + aSpan;
    betaLo = bestBeta / 2.5;
    betaHi = bestBeta * 2.5;
  }

  // Threshold: x where the curve crosses the target proportion.
  double? threshold;
  final sigma = (targetProportion - guessRate) / (1 - guessRate - lapseRate);
  if (sigma > 0 && sigma < 1) {
    threshold = bestAlpha + math.log(sigma / (1 - sigma)) / bestBeta;
  }

  return LogisticFit(
    alpha: bestAlpha,
    beta: bestBeta,
    gamma: guessRate,
    lambda: lapseRate,
    targetProportion: targetProportion,
    threshold: threshold,
    nTrials: x.length,
  );
}

/// Two-way ANOVA mean squares for an n×k score matrix: (MSR, MSC, MSE).
/// Returns null when the matrix is smaller than 2×2 or ragged.
(double, double, double)? _iccMeanSquares(List<List<double>> data) {
  final n = data.length;
  if (n < 2) return null;
  final k = data.first.length;
  if (k < 2) return null;
  for (final row in data) {
    if (row.length != k) return null;
  }
  var grand = 0.0;
  for (final row in data) {
    for (final v in row) {
      grand += v;
    }
  }
  grand /= n * k;
  final rowMeans = <double>[
    for (final row in data) row.reduce((a, b) => a + b) / k,
  ];
  final colMeans = List<double>.filled(k, 0);
  for (final row in data) {
    for (var j = 0; j < k; j++) {
      colMeans[j] += row[j] / n;
    }
  }
  var ssr = 0.0, ssc = 0.0, sse = 0.0;
  for (var i = 0; i < n; i++) {
    ssr += (rowMeans[i] - grand) * (rowMeans[i] - grand);
    for (var j = 0; j < k; j++) {
      final resid = data[i][j] - rowMeans[i] - colMeans[j] + grand;
      sse += resid * resid;
    }
  }
  for (var j = 0; j < k; j++) {
    ssc += (colMeans[j] - grand) * (colMeans[j] - grand);
  }
  final msr = k * ssr / (n - 1);
  final msc = n * ssc / (k - 1);
  final mse = sse / ((n - 1) * (k - 1));
  return (msr, msc, mse);
}

/// ICC(2,1): two-way random effects, absolute agreement, single measurement
/// (Shrout & Fleiss, 1979) — the standard test–retest reliability coefficient
/// when systematic session-to-session shifts should count against agreement.
///
/// [data] is an n-subjects × k-occasions matrix (each row one subject's scores
/// on the same test, in occasion order). Returns null for a matrix smaller
/// than 2×2, ragged rows, or zero variance (nothing to apportion).
double? icc21(List<List<double>> data) {
  final ms = _iccMeanSquares(data);
  if (ms == null) return null;
  final (msr, msc, mse) = ms;
  final n = data.length, k = data.first.length;
  final denom = msr + (k - 1) * mse + k * (msc - mse) / n;
  if (denom.abs() < 1e-12) return null;
  return (msr - mse) / denom;
}

/// ICC(3,1): two-way mixed effects, consistency, single measurement —
/// reliability when systematic occasion shifts (e.g. practice effects) are
/// excluded and only relative agreement matters. Same input/degeneracy rules
/// as [icc21].
double? icc31(List<List<double>> data) {
  final ms = _iccMeanSquares(data);
  if (ms == null) return null;
  final (msr, _, mse) = ms;
  final k = data.first.length;
  final denom = msr + (k - 1) * mse;
  if (denom.abs() < 1e-12) return null;
  return (msr - mse) / denom;
}
